mopidy/m3u/playlists.py

```diff
@@ -1,117 +1,117 @@
+        self._default_encoding = ext_config['default_encoding']
+        self._default_extension = ext_config['default_extension']
 
-    def get_items(self, uri):
-        playlist = self._playlists.get(uri)
-        if playlist is None:
-            return None
-        return [Ref.track(uri=t.uri, name=t.name) for t in playlist.tracks]
+    def as_list(self):
+        result = []
+        for entry in os.listdir(self._playlists_dir):
+            if not entry.endswith((b'.m3u', b'.m3u8')):
+                continue
+            elif not os.path.isfile(self._abspath(entry)):
+                continue
+            else:
+                result.append(translator.path_to_ref(entry))
+        result.sort(key=operator.attrgetter('name'))
+        return result
 
     def create(self, name):
-        playlist = self._save_m3u(Playlist(name=name))
-        self._playlists[playlist.uri] = playlist
-        logger.info('Created playlist %s', playlist.uri)
-        return playlist
+        path = translator.path_from_name(name.strip(), self._default_extension)
+        try:
+            with self._open(path, 'w'):
+                pass
+            mtime = os.path.getmtime(self._abspath(path))
+        except EnvironmentError as e:
+            log_environment_error('Error creating playlist %s' % name, e)
+        else:
+            return translator.playlist(path, [], mtime)
 
     def delete(self, uri):
-        if uri in self._playlists:
-            path = translator.playlist_uri_to_path(uri, self._playlists_dir)
-            if os.path.exists(path):
-                os.remove(path)
-            else:
-                logger.warning(
-                    'Trying to delete missing playlist file %s', path)
-            del self._playlists[uri]
-            logger.info('Deleted playlist %s', uri)
+        path = translator.uri_to_path(uri)
+        try:
+            os.remove(self._abspath(path))
+        except EnvironmentError as e:
+            log_environment_error('Error deleting playlist %s' % uri, e)
+
+    def get_items(self, uri):
+        path = translator.uri_to_path(uri)
+        try:
+            with self._open(path, 'r') as fp:
+                items = translator.load_items(fp, self._playlists_dir)
+        except EnvironmentError as e:
+            log_environment_error('Error reading playlist %s' % uri, e)
         else:
-            logger.warning('Trying to delete unknown playlist %s', uri)
+            return items
 
     def lookup(self, uri):
-        return self._playlists.get(uri)
+        path = translator.uri_to_path(uri)
+        try:
+            with self._open(path, 'r') as fp:
+                items = translator.load_items(fp, self._playlists_dir)
+            mtime = os.path.getmtime(self._abspath(path))
+        except EnvironmentError as e:
+            log_environment_error('Error reading playlist %s' % uri, e)
+        else:
+            return translator.playlist(path, items, mtime)
 
     def refresh(self):
```

Comment:
Well, I think getting rid of the `self._playlists` dict and following the `file` backend is probably the most radical change in this PR, so I'm quite curious how clients will react to that ;-)
However, triggering a `playlist_changed` event in `lookup` or `as_list` or whatever doesn't make *that* much sense, IMHO. The client triggering the `lookup()` or whatever will notice the change anyway, so there would be only a benefit if *another* client accidentaly triggers this, AFAICS.
If we were watching the `playlist_dir` using `inotify`, for example, triggering `playlist_changed` events *would* make sense. And when modifying playlists via the Mopidy API, `playlist_changed` events *should* be sent, of course.